src/sentry/middleware/integrations/classifications.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import annotations
+
+import abc
+import logging
+import re
+from typing import TYPE_CHECKING, Callable, Type
+
+from django.http import HttpRequest, HttpResponse
+from pyparsing import Mapping
+
+from .parsers import (
+    BitbucketRequestParser,
+    BitbucketServerRequestParser,
+    GithubEnterpriseRequestParser,
+    GithubRequestParser,
+    GitlabRequestParser,
+    JiraRequestParser,
+    JiraServerRequestParser,
+    MsTeamsRequestParser,
+    PluginRequestParser,
+    SlackRequestParser,
+    VstsRequestParser,
+)
+
+if TYPE_CHECKING:
+    from .parsers.base import BaseRequestParser
+
+
+class BaseClassification(abc.ABC):
+    def __init__(self, response_handler: Callable[[], HttpResponse]) -> None:
+        self.response_handler = response_handler
+
+    def should_operate(self, request: HttpRequest) -> bool:
+        """
+        Determines whether the classification should act on request.
+        Middleware will return self.get_response() if this function returns True.
+        """
+        raise NotImplementedError
+
+    def get_response(self, request: HttpRequest) -> HttpResponse:
+        """Parse the request and return a response."""
+        raise NotImplementedError
+
+
+class PluginClassification(BaseClassification):
+    def should_operate(self, request: HttpRequest) -> bool:
+        rp = PluginRequestParser(request=request, response_handler=self.response_handler)
+        return rp.should_operate()
+
+    def get_response(self, request: HttpRequest) -> HttpResponse:
+        rp = PluginRequestParser(request=request, response_handler=self.response_handler)
+        return rp.get_response()
+
+
+class IntegrationClassification(BaseClassification):
```

Comment:
All of this code has been refactored out of the `IntegrationControlMiddleware` with no significant changes.